File: client/uiwails/services/connection.go
//go:build !(linux && 386)

package services

import (
	"context"
	"fmt"
	"time"

	log "github.com/sirupsen/logrus"

	"github.com/netbirdio/netbird/client/proto"
)

// ConnectionService exposes connect/disconnect/status operations to the Wails frontend.
type ConnectionService struct {
	grpcClient GRPCClientIface
}

// NewConnectionService creates a new ConnectionService.
func NewConnectionService(g GRPCClientIface) *ConnectionService {
	return &ConnectionService{grpcClient: g}
}

// GetStatus returns the current daemon status.
func (s *ConnectionService) GetStatus() (*StatusInfo, error) {
	conn, err := s.grpcClient.GetClient(3 * time.Second)
	if err != nil {
		log.Debugf("GetStatus: failed to get gRPC client: %v", err)
		return nil, fmt.Errorf("get client: %w", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	resp, err := conn.Status(ctx, &proto.StatusRequest{GetFullPeerStatus: true})
	if err != nil {
		log.Warnf("GetStatus: status RPC failed: %v", err)
		return nil, fmt.Errorf("status rpc: %w", err)
	}

	log.Debugf("GetStatus: daemon responded status=%q daemonVersion=%q fullStatus=%v",
		resp.Status, resp.DaemonVersion, resp.FullStatus != nil)

	info := &StatusInfo{
		Status: resp.Status,
	}

	if resp.FullStatus != nil && resp.FullStatus.LocalPeerState != nil {
		lp := resp.FullStatus.LocalPeerState
		info.IP = lp.GetIP()
		info.PublicKey = lp.GetPubKey()
		info.Fqdn = lp.GetFqdn()
		log.Debugf("GetStatus: localPeer ip=%q fqdn=%q pubKey=%q", info.IP, info.Fqdn, info.PublicKey)
	} else if resp.FullStatus == nil {
		log.Warnf("GetStatus: fullStatus is nil — daemon may not support full status or request flag was not set")
	} else {
		log.Debugf("GetStatus: fullStatus present but LocalPeerState is nil")
	}

	if resp.FullStatus != nil {
		info.ConnectedPeers = len(resp.FullStatus.GetPeers())
		log.Debugf("GetStatus: connectedPeers=%d", info.ConnectedPeers)
	}

	return info, nil
}

// Connect sends an Up request to the daemon.
func (s *ConnectionService) Connect() error {
	conn, err := s.grpcClient.GetClient(3 * time.Second)
	if err != nil {
		return fmt.Errorf("get client: %w", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	if _, err := conn.Up(ctx, &proto.UpRequest{}); err != nil {
		log.Errorf("Up rpc failed: %v", err)
		return fmt.Errorf("connect: %w", err)
	}

	return nil
}

// Disconnect sends a Down request to the daemon.
func (s *ConnectionService) Disconnect() error {
	conn, err := s.grpcClient.GetClient(3 * time.Second)
	if err != nil {
		return fmt.Errorf("get client: %w", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	if _, err := conn.Down(ctx, &proto.DownRequest{}); err != nil {
		log.Errorf("Down rpc failed: %v", err)
		return fmt.Errorf("disconnect: %w", err)
	}

	return nil
}

// StatusInfo holds simplified status information for the frontend.
type StatusInfo struct {
	Status         string `json:"status"`
	IP             string `json:"ip"`
	PublicKey      string `json:"publicKey"`
	Fqdn           string `json:"fqdn"`
	ConnectedPeers int    `json:"connectedPeers"`
}

File: client/uitauri/src-tauri/src/commands/connection.rs
use serde::Serialize;
use tauri::State;

use crate::proto;
use crate::state::AppState;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusInfo {
    pub status: String,
    pub ip: String,
    pub public_key: String,
    pub fqdn: String,
    pub connected_peers: usize,
}

#[tauri::command]
pub async fn get_status(state: State<'_, AppState>) -> Result<StatusInfo, String> {
    let mut client = state.grpc.get_client().await?;
    let resp = client
        .status(proto::StatusRequest {
            get_full_peer_status: true,
            should_run_probes: false,
            wait_for_ready: None,
        })
        .await
        .map_err(|e| format!("status rpc: {}", e))?
        .into_inner();

    let mut info = StatusInfo {
        status: resp.status,
        ip: String::new(),
        public_key: String::new(),
        fqdn: String::new(),
        connected_peers: 0,
    };

    if let Some(ref full) = resp.full_status {
        if let Some(ref lp) = full.local_peer_state {
            info.ip = lp.ip.clone();
            info.public_key = lp.pub_key.clone();
            info.fqdn = lp.fqdn.clone();
        }
        info.connected_peers = full.peers.len();
    }

    Ok(info)
}

#[tauri::command]
pub async fn connect(state: State<'_, AppState>) -> Result<(), String> {
    let mut client = state.grpc.get_client().await?;
    client
        .up(proto::UpRequest {
            profile_name: None,
            username: None,
            auto_update: None,
        })
        .await
        .map_err(|e| format!("connect: {}", e))?;
    Ok(())
}

#[tauri::command]
pub async fn disconnect(state: State<'_, AppState>) -> Result<(), String> {
    let mut client = state.grpc.get_client().await?;
    client
        .down(proto::DownRequest {})
        .await
        .map_err(|e| format!("disconnect: {}", e))?;
    Ok(())
}
